### PR TITLE
[Feature] add hash function for xx_hash3_64 

### PR DESCRIPTION
**Pull request by goldenbean**
xx_hash3_64 has much better performance than murmur_hash3_32 by using AVX2 instruction according their benchmark, and has state-of-art hash quality which is broadly integrated in many software.

Reference: https://github.com/Cyan4973/xxHash

## What type of PR is this:
- [ ] BugFix
- [X] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by imay** [comment]
Did you test the performance based on this PR compared to murmur_hash3_32?

**Comment by goldenbean** [comment]
> Did you test the performance based on this PR compared to murmur_hash3_32?

Since the implementation is quite simple and straight forward, I'm referred to the official benchmarks among the popular algo in the code "be/src/util/xxhash.h" as following: 

* Benchmarks
 * ---
 * The reference system uses an Intel i7-9700K CPU, and runs Ubuntu x64 20.04.
 * The open source benchmark program is compiled with clang v10.0 using -O3 flag.
 *
 * | Hash Name            | ISA ext | Width | Large Data Speed | Small Data Velocity |
 * | -------------------- | ------- | ----: | ---------------: | ------------------: |
 * | XXH3_64bits()        | @b AVX2 |    64 |        59.4 GB/s |               133.1 |
 * | MeowHash             | AES-NI  |   128 |        58.2 GB/s |                52.5 |
 * | XXH3_128bits()       | @b AVX2 |   128 |        57.9 GB/s |               118.1 |
 * | CLHash               | PCLMUL  |    64 |        37.1 GB/s |                58.1 |
 * | XXH3_64bits()        | @b SSE2 |    64 |        31.5 GB/s |               133.1 |
 * | XXH3_128bits()       | @b SSE2 |   128 |        29.6 GB/s |               118.1 |
 * | RAM sequential read  |         |   N/A |        28.0 GB/s |                 N/A |
 * | ahash                | AES-NI  |    64 |        22.5 GB/s |               107.2 |
 * | City64               |         |    64 |        22.0 GB/s |                76.6 |
 * | T1ha2                |         |    64 |        22.0 GB/s |                99.0 |
 * | City128              |         |   128 |        21.7 GB/s |                57.7 |
 * | FarmHash             | AES-NI  |    64 |        21.3 GB/s |                71.9 |
 * | XXH64()              |         |    64 |        19.4 GB/s |                71.0 |
 * | SpookyHash           |         |    64 |        19.3 GB/s |                53.2 |
 * | Mum                  |         |    64 |        18.0 GB/s |                67.0 |
 * | CRC32C               | SSE4.2  |    32 |        13.0 GB/s |                57.9 |
 * | XXH32()              |         |    32 |         9.7 GB/s |                71.9 |
 * | City32               |         |    32 |         9.1 GB/s |                66.0 |
 * | Blake3*              | @b AVX2 |   256 |         4.4 GB/s |                 8.1 |
 * | Murmur3              |         |    32 |         3.9 GB/s |                56.1 |
 * | SipHash*             |         |    64 |         3.0 GB/s |                43.2 |
 * | Blake3*              | @b SSE2 |   256 |         2.4 GB/s |                 8.1 |
 * | HighwayHash          |         |    64 |         1.4 GB/s |                 6.0 |
 * | FNV64                |         |    64 |         1.2 GB/s |                62.7 |
 * | Blake2*              |         |   256 |         1.1 GB/s |                 5.1 |
 * | SHA1*                |         |   160 |         0.8 GB/s |                 5.6 |
 * | MD5*                 |         |   128 |         0.6 GB/s |                 7.8 |

**Comment by imay** [comment]
I have seen the benchmark result. It is better to do tests based on this PR to make sure better performance can even exist in starrocks.

**Comment by goldenbean** [comment]
> I have seen the benchmark result. It is better to do tests based on this PR to make sure better performance can even exist in starrocks.

Got you

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by imay** [comment]
@goldenbean can you share the performance results of this new function and murmur_hash3_32?

**Comment by goldenbean** [comment]
BENCHMARK


Run on (12 X 2100 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x6)
  L1 Instruction 32 KiB (x6)
  L2 Unified 512 KiB (x6)
  L3 Unified 4096 KiB (x2)
Load Average: 3.21, 5.49, 7.78
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
---------------------------------------------------------------------------------------------
Benchmark                                                   Time             CPU   Iterations
---------------------------------------------------------------------------------------------
BM_HashFunctions_Eval/1/10/1/iterations:10000             354 ns          352 ns        10000
BM_HashFunctions_Eval/1/10/0/iterations:10000             333 ns          332 ns        10000
BM_HashFunctions_Eval/1/100/1/iterations:10000           1506 ns         1503 ns        10000
BM_HashFunctions_Eval/1/100/0/iterations:10000           1269 ns         1268 ns        10000
BM_HashFunctions_Eval/1/10000/1/iterations:10000       225514 ns       225380 ns        10000
BM_HashFunctions_Eval/1/10000/0/iterations:10000       156346 ns       156262 ns        10000
BM_HashFunctions_Eval/1/1000000/1/iterations:10000   22690231 ns     22659365 ns        10000
BM_HashFunctions_Eval/1/1000000/0/iterations:10000   19238351 ns     19210546 ns        10000

**Comment by imay** [comment]
Form the benchmark, we can see the improvement. But the speed seems not fast enough as Reference: https://github.com/Cyan4973/xxHash. do you know the reason?

**Comment by goldenbean** [comment]
@imay please let me do more tests on commercial server on Monday.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by goldenbean** [comment]
-------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations
-------------------------------------------------------------------------------------------
BM_HashFunctions_Eval/10/1/iterations:10000             722 ns          722 ns        10000
BM_HashFunctions_Eval/10/0/iterations:10000             609 ns          609 ns        10000
BM_HashFunctions_Eval/100/1/iterations:10000           3720 ns         3720 ns        10000
BM_HashFunctions_Eval/100/0/iterations:10000           2534 ns         2534 ns        10000
BM_HashFunctions_Eval/10000/1/iterations:10000       411227 ns       411221 ns        10000
BM_HashFunctions_Eval/10000/0/iterations:10000       220210 ns       220207 ns        10000
BM_HashFunctions_Eval/1000000/1/iterations:10000   41261672 ns     41261079 ns        10000
BM_HashFunctions_Eval/1000000/0/iterations:10000   23734210 ns     23733623 ns        10000

Hardware:
CPU: 64 core  Intel(R) Xeon(R) CPU E5-2683 v4 @ 2.10GHz
MEM: 256 GB

Run on (64 X 2593.98 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x32)
  L1 Instruction 32 KiB (x32)
  L2 Unified 256 KiB (x32)
  L3 Unified 40960 KiB (x2)
Load Average: 0.00, 0.02, 0.00

File: be/src/exprs/hash_functions.h [hunk 1]
```diff
@@ -58,4 +64,33 @@ inline StatusOr<ColumnPtr> HashFunctions::murmur_hash3_32(FunctionContext* conte
...
+    std::vector<ColumnViewer<TYPE_VARCHAR>> viewers;
+
+    viewers.reserve(columns.size());
+    for (const auto& column : columns) {
+        viewers.emplace_back(column);
+    }
+
+    size_t size = columns[0]->size();
+    ColumnBuilder<TYPE_BIGINT> builder(size);
+    for (int row = 0; row < size; ++row) {
```
Comment (satanson) [inline be/src/exprs/hash_functions.h:77]: suggest to transform this row-priority two-layer loops into column-priority loops, compiler shall try best to auto vectorize column-priority implement.

Comment (goldenbean) [reply be/src/exprs/hash_functions.h:77]: nice catch ! I will do it soon

**Comment by imay** [comment]
@goldenbean better to rename it as `xx_hash3_64` which will be consistent with `murmur_hash3_32 `

**Comment by goldenbean** [comment]
> @goldenbean better to rename it as `xx_hash3_64` which will be consistent with `murmur_hash3_32 `

ok

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by satanson — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Comment by rcauble** [comment]
Hi, can this be backported to 3.1 branch, please?

**Comment by imay** [comment]
@mergify backport branch-3.1

**Pull request merged**